### PR TITLE
test: Handle functional test disk-full error

### LINKED ISSUE #23099: Handle disk-full more gracefully in functional tests
If you run the tests with too little disk space in  `/tmp` partition, e.g.
```bash
sudo mount -t tmpfs -o size=1M none /tmp
test/functional/test_runner.py
```
you get a huge log output with a lot of `-- Logging error --`, with the tests failing one by one, where it's not immediately clear why the tests are failing. It might be useful to exit immediately with a more specific message when even writing to the logs fails, and/or do an early check if the disk space need in the file system where the temporary files go can be estimated.

### PR DESCRIPTION
**Pull request by BrandonOdiwuor**
Fixes: https://github.com/bitcoin/bitcoin/issues/23099

Handle disk-full more gracefully in functional tests

### CODE REVIEW

File: test/functional/test_runner.py [hunk 1]
```diff
@@ -30,6 +30,9 @@
 
 os.environ["REQUIRE_WALLET_TYPE_SET"] = "1"
 
+# Should fail testing and immediately stop if free space falls below.
+MIN_FREE_SPACE = (50 + 16) * 1024 * 1024
```
[the remark this answers is not in the document]
Comment (BrandonOdiwuor) [reply test/functional/test_runner.py:34]: I based the values based on the discussions on  https://github.com/bitcoin/bitcoin/pull/23100#issuecomment-941115724 and  https://github.com/bitcoin/bitcoin/pull/23100#discussion_r716825510 and 
Screenshot from 2024-01-28 15-08-25

File: test/functional/test_runner.py [hunk 2]
```diff
@@ -639,6 +642,11 @@ def run_tests(*, test_list, src_dir, build_dir, tmpdir, jobs=1, enable_coverage=
                     logging.debug("Early exiting after test failure")
                     break
 
+                testdir_disk_usage_stats = shutil.disk_usage(testdir)
+                if (testdir_disk_usage_stats.free < MIN_FREE_SPACE):
+                    logging.debug("Not space enough on device: %s" % testdir)
```
[the remark this answers is not in the document]
Comment (BrandonOdiwuor) [reply test/functional/test_runner.py]: fixed

File: test/functional/test_runner.py [hunk 3]
```diff
@@ -639,6 +642,11 @@ def run_tests(*, test_list, src_dir, build_dir, tmpdir, jobs=1, enable_coverage=
                     logging.debug("Early exiting after test failure")
                     break
 
+                testdir_disk_usage_stats = shutil.disk_usage(testdir)
```
[the remark this answers is not in the document]
Comment (BrandonOdiwuor) [reply test/functional/test_runner.py]: moved the check up `run_checks(...)`

**Comment by kevkevinpal** [comment]
Concept ACK [ebb842a](https://github.com/bitcoin/bitcoin/pull/29335/commits/ebb842ae4fbb20d2933e6852afd055bedeacf8c2)

File: test/functional/test_runner.py [hunk 4]
```diff
@@ -574,6 +577,11 @@ def run_tests(*, test_list, src_dir, build_dir, tmpdir, jobs=1, enable_coverage=
     if not result.wasSuccessful():
         sys.exit("Early exiting after failure in TestFramework unit tests")
 
+    # Exit the test if there is not enough space on the testing dir
+    testdir_disk_usage_stats = shutil.disk_usage(tmpdir)
+    if (testdir_disk_usage_stats.free < MIN_FREE_SPACE):
+        sys.exit(f"Tests aborted: Insufficient space available on the device. Directory: {tmpdir}")
```
Comment (kevkevinpal) [inline test/functional/test_runner.py:583]: I think we can reword this to 
`Tests aborted: Insufficient space available in directory: {tmpdir}`

Since the device may have space but the `tmpdir` we are using might have insufficient space

File: test/functional/test_runner.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (BrandonOdiwuor) [reply test/functional/test_runner.py:34]: The total size of the generated testdir after running `./test_runner.py --nocleanup` is `11GB` on my system
Screenshot from 2024-01-30 13-58-06

See the list of the generated folders and their sizes on the attached file below:
`sudo du -h /tmp/test_runner_₿_🏃_20240130_125641/ > ~/tmp-test_runner_₿_🏃_20240130_125641-test-files.txt`

[tmp-test_runner_₿_🏃_20240130_125641-test-files.txt](https://github.com/bitcoin/bitcoin/files/14096722/tmp-test_runner_._._20240130_125641-test-files.txt)

**Comment by epiccurious** [comment]
utACK

**Comment by epiccurious** [comment]
Right now, the test aborts for low disk space after running the Unit Tests for Test Framework Modules.

Instead, does it make sense to abort for low disk space before?

**Comment by BrandonOdiwuor** [comment]
@epiccurious moved aborting the tests before the Unit Tests for Test Framework Modules.

File: test/functional/test_runner.py [hunk 5]
```diff
@@ -559,6 +562,11 @@ def run_tests(*, test_list, src_dir, build_dir, tmpdir, jobs=1, enable_coverage=
     if os.path.isdir(cache_dir):
         print("%sWARNING!%s There is a cache directory here: %s. If tests fail unexpectedly, try deleting the cache directory." % (BOLD[1], BOLD[0], cache_dir))
 
+    # Exit the test if there is not enough space on the testing dir
+    testdir_disk_usage_stats = shutil.disk_usage(tmpdir)
+    if (testdir_disk_usage_stats.free < MIN_FREE_SPACE):
```
[the remark this answers is not in the document]
Comment (BrandonOdiwuor) [reply test/functional/test_runner.py]: fixed

**Review by maflcko — APPROVED** [review]
lgtm

File: test/functional/test_runner.py [hunk 6]
```diff
@@ -559,6 +562,11 @@ def run_tests(*, test_list, src_dir, build_dir, tmpdir, jobs=1, enable_coverage=
     if os.path.isdir(cache_dir):
         print("%sWARNING!%s There is a cache directory here: %s. If tests fail unexpectedly, try deleting the cache directory." % (BOLD[1], BOLD[0], cache_dir))
 
+    # Exit the test if there is not enough space on the testing dir
+    testdir_disk_usage_stats = shutil.disk_usage(tmpdir)
+    if testdir_disk_usage_stats.free < MIN_FREE_SPACE:
```
[the remark this answers is not in the document]
Comment (BrandonOdiwuor) [reply test/functional/test_runner.py]: fixed

File: test/functional/test_runner.py [hunk 7]
```diff
@@ -655,7 +662,7 @@ def run_tests(*, test_list, src_dir, build_dir, tmpdir, jobs=1, enable_coverage=
 
     all_passed = all_passed and coverage_passed
 
-    # Clean up dangling processes if any. This may only happen with --failfast option.
+    # Clean up dangling processes if any. This may only happen with --failfast option or not enough space on test directory.
```
[the remark this answers is not in the document]
Comment (BrandonOdiwuor) [reply test/functional/test_runner.py]: fixed

**Comment by kevkevinpal** [comment]
tested ACK [16baf86](https://github.com/bitcoin/bitcoin/pull/29335/commits/16baf8651d17ac1e0c7d7fd3526c19ad018ceb9c)

This may be out of scope for this PR but I have filled up an SD card almost full and when I run the following,

## I get the correct response
`./test/functional/test_runner.py -t /Volumes/32GBSDCARD/`
```
Temporary test directory at /Volumes/32GBSDCARD//test_runner_₿_🏃_20240222_215044
Tests aborted: Insufficient space available in directory: /Volumes/32GBSDCARD//test_runner_₿_🏃_20240222_215044
```

-----

## But when I run an individual test for example
`./test/functional/mempool_accept.py --tmpdir /Volumes/32GBSDCARD/`
I get the following error
```
Traceback (most recent call last):
  File "/Users/kevkevin/DEVDIR/bitcoin/test/functional/test_framework/test_framework.py", line 130, in main
    self.setup()
  File "/Users/kevkevin/DEVDIR/bitcoin/test/functional/test_framework/test_framework.py", line 266, in setup
    os.makedirs(self.options.tmpdir, exist_ok=False)
  File "<frozen os>", line 225, in makedirs
FileExistsError: [Errno 17] File exists: '/Volumes/32GBSDCARD'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/kevkevin/DEVDIR/bitcoin/test/functional/test_framework/test_framework.py", line 148, in main
    self.log.exception("Unexpected exception caught during testing")
    ^^^^^^^^
AttributeError: 'MempoolAcceptanceTest' object has no attribute 'log'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/kevkevin/DEVDIR/bitcoin/./test/functional/mempool_accept.py", line 376, in <module>
    MempoolAcceptanceTest().main()
  File "/Users/kevkevin/DEVDIR/bitcoin/test/functional/test_framework/test_framework.py", line 154, in main
    exit_code = self.shutdown()
                ^^^^^^^^^^^^^^^
  File "/Users/kevkevin/DEVDIR/bitcoin/test/functional/test_framework/test_framework.py", line 304, in shutdown
    if self.success == TestStatus.FAILED and self.options.pdbonfailure:
       ^^^^^^^^^^^^
AttributeError: 'MempoolAcceptanceTest' object has no attribute 'success'
```

-----

## The disk that I used
`df -h`
`/dev/disk8s1      30Gi    29Gi   160Mi   100%    2.5k  1.6M    0%   /Volumes/32GBSDCARD`

File: test/functional/test_runner.py [hunk 8]
```diff
@@ -30,6 +30,9 @@
 
 os.environ["REQUIRE_WALLET_TYPE_SET"] = "1"
 
+# Should fail testing and immediately stop if free space falls below.
+MIN_FREE_SPACE = 1250 * 1024 * 1024
```
Comment (cbergqvist) [inline test/functional/test_runner.py:34]: 2024-02-23 14-08-26_Disk_usage

Thanks for the monitoring script you provided earlier @BrandonOdiwuor! Was able to get disk usage almost at 5GB using:
```
$ ./test/functional/test_runner.py --cachedir=/mnt/tmp/cache --tmpdir=/mnt/tmp --extended --jobs=16
```
So maybe some heuristic could be applied when running many jobs in parallel?
Default number of jobs is 4, so maybe something like
```python
MIN_FREE_SPACE = min(jobs, 4) * 320 * 1024 * 1024
```
That way we get `1280 * 1024 * 1024` for 4 jobs or less, and `5120 * 1024 * 1024` for 16 jobs.

File: test/functional/test_runner.py [hunk 9]
```diff
@@ -559,6 +562,10 @@ def run_tests(*, test_list, src_dir, build_dir, tmpdir, jobs=1, enable_coverage=
     if os.path.isdir(cache_dir):
         print("%sWARNING!%s There is a cache directory here: %s. If tests fail unexpectedly, try deleting the cache directory." % (BOLD[1], BOLD[0], cache_dir))
 
+    # Exit the test if there is not enough space on the testing dir
+    if shutil.disk_usage(tmpdir).free < MIN_FREE_SPACE:
+        sys.exit(f"Tests aborted: Insufficient space available in directory: {tmpdir}")
```
Comment (cbergqvist) [inline test/functional/test_runner.py:567]: Since the `MIN_FREE_SPACE` value is a bit of a guess, it might make sense to either make this into a VERY LOUD non-fatal warning, or add an arg like `--non-fatal-disk-space-check` to make it into a warning. That way people running in a bit resource-constrained environments and/or only interested in a specific low-disk-usage test don't need to edit the `MIN_FREE_SPACE` to be able to run their tests.

**Review by cbergqvist — CHANGES_REQUESTED** [review]

File: test/functional/test_runner.py (hunk 9 above)
Comment (BrandonOdiwuor) [reply test/functional/test_runner.py]: @maflcko I have converted this to a warning as suggested

**Review by cbergqvist — APPROVED** [review]
ACK 951a004!

**Review by marcofleon — COMMENTED** [review]
I managed to create a disk image with available space just above `MIN_FREE_SPACE` such that the warning is bypassed when running `test_runner.py` but one of the tests (`feature_signet.py`) fails with error:
```bash
    raise AssertionError("Unexpected stderr {} != {}".format(stderr, expected_stderr))
AssertionError: Unexpected stderr Warning: Disk space for "/Volumes/Untitled/test_runner_₿_🏃_20240227_223648/feature_signet_91/node2/signet/blocks" may not accommodate the block files. Approximately 1 GB of data will be stored in this directory. != 
```
All of the tests pass for me when there's about 1.35 GB of space available, so `1280*1024*1024` would probably work better.

However, we've already acknowledged that this variable is sort of just a guess, so I'm not sure how critical it is to get exactly right. I'm also not sure how different these tests' disk space usage is across systems/machines. The fact that it's shown only as a warning makes sense then.

nit: I think the warning could be louder. The check above uses bold in its warning. Maybe this one can follow suit?

**Comment by davidgumberg** [comment]
I think it would be good to stop the test runner once we actually run into an out of disk space error

Something like this:

```diff
 while i < test_count:
     if failfast and not all_passed:
         break
     for test_result, testdir, stdout, stderr, skip_reason in job_queue.get_next():
         test_results.append(test_result)
         i += 1
         done_str = "{}/{} - {}{}{}".format(i, test_count, BOLD[1], test_result.name, BOLD[0])
         if test_result.status == "Passed":
             logging.debug("%s passed, Duration: %s s" % (done_str, test_result.time))
         elif test_result.status == "Skipped":
             logging.debug(f"{done_str} skipped ({skip_reason})")
         else:
             all_passed = False
             print("%s failed, Duration: %s s\n" % (done_str, test_result.time))
             print(BOLD[1] + 'stdout:\n' + BOLD[0] + stdout + '\n')
             print(BOLD[1] + 'stderr:\n' + BOLD[0] + stderr + '\n')
             if combined_logs_len and os.path.isdir(testdir):
                 pass #[omitted]

+       if stderr.strip().split('\n')[-1] == "OSError: [Errno 28] No space left on device":
+           logging.debug(f"Early exiting after test failure due to disk running out of space in {tmpdir}")
+           break
+
        if failfast:
            logging.debug("Early exiting after test failure")
            break
 ```

We could also avoid the question of picking a good number for warning if we used this approach without any warning, while still solving the underlying issue of stopping the test runner from continuing to dump test failures and errors that are not obviously related to being out of disk space, and having an opportunity to communicate the reason for failure to the user.

I am not opposed to the warning message, but if we keep it I think it should
1) use a number that is closer to the upper bound of space that might be used
2) use language more like: `there may be insufficient free space in {tmpdir} to run the bitcoin functional test suite. Running the test suite with fewer than {amt} of free space might cause tests to fail.`

**Comment by kevkevinpal** [comment]
reACK 951a00440539ba1faba62c49c9cf0a7d8c950769

Tested with `--failfast` and normally and both seem to be working properly with the correct error message

**Comment by BrandonOdiwuor** [comment]
cc @maflcko could you have another look at this

**Review by achow101 — COMMENTED** [review]
I agree that if a test fails out of disk space, then we should stop running any further tests as well.

I think it would be useful for each test to also have a disk space check, and that could even have its own exit code that the test runner can use to detect when a test failed because of low disk space.

**Comment by BrandonOdiwuor** [comment]
@achow101 @davidgumberg would it be a good idea to do both i.e include the check with a warning at the beginning and also check after every test is run if any fails then stop running any further test?

**Comment by achow101** [comment]
> @achow101 @davidgumberg would it be a good idea to do both i.e include the check with a warning at the beginning and also check after every test is run if any fails then stop running any further test?

Yes, I think so.

File: test/functional/test_runner.py [hunk 10]
```diff
@@ -30,6 +30,9 @@
 
 os.environ["REQUIRE_WALLET_TYPE_SET"] = "1"
 
+# Minimum amount of space to run the tests.
+MIN_FREE_SPACE = 1250 * 1024 * 1024
```
[the remark this answers is not in the document]
Comment (BrandonOdiwuor) [reply test/functional/test_runner.py]: @achow101 could you elaborate more on this
> Also, if --nocleanup is provided, we'd probably want to warn at the size where nothing is cleaned up.

[the remark this answers is not in the document]
Comment (BrandonOdiwuor) [reply test/functional/test_runner.py]: I have updated the PR as suggested:

> When a high enough parallelism is used, this is not enough.

>It may be better to scale this based on the number of jobs that will be run in parallel. Could put a lower bound of 1.1 GB for feature_block which is an outlier, then start increasing by 100 MB for each additional job that will be run.

[the remark this answers is not in the document]
Comment (achow101) [reply test/functional/test_runner.py]: > @achow101 could you elaborate more on this
> 
> > Also, if --nocleanup is provided, we'd probably want to warn at the size where nothing is cleaned up.

`--nocleanup` requires about 12 GB, so if that option is passed, the check should be for that amount.

[the remark this answers is not in the document]
Comment (BrandonOdiwuor) [reply test/functional/test_runner.py]: updated the PR to include the check for minimum space required for `--nocleanup`

**Comment by itornaza** [comment]
tested ACK 2f7fcd5fe7289f7ad3efd566dbbd725264654f8f

The user is notified when appropriate that there might be insufficient free space to run the tests. Moreover, if a test fails due to insufficient space, the functional test runner halts. Also, there is a relevant warning when the `--nocleanup` flag is used and hence the free disk requirements get bigger.

Note: All tests run on macos 14.4.1 (23E224)

## Standard tests

- Did a code review and checked that there is adequate space on the file system to run the tests
- Configured with `--with-incompatible-bdb` and `--enable-suppress-external-warnings`
- Run unit tests with `make check` and all tests pass
- Run all functional tests with no extra flags and all tests pass and no warnings about free space
- Run all functional tests with `--extended` and all tests pass

## Manual tests

Used `-j 13` for the number of jobs to be run in parallel since the M2 Pro Max chip supports them. Calculated the required space to run the tests with the 13 parallel jobs as described in the pr:

`1.1 * 1024 + (13 - 1) * 100 MB = 2326 MB`

and reated an image at `/Volumes/tmp` with the a little less space set to 2.2 GB, with the `df -h /Volumes/tmp` command returning: 

```
Filesystem      Size    Used   Avail Capacity iused ifree %iused  Mounted on
/dev/disk4s2   2.2Gi    16Mi   2.2Gi     1%       4  4.3G    0%   /Volumes/tmp
```

Running the functional tests with `./test/functional/test_runner.py -j13 -t /Volumes/tmp` I get the warning up front and all tests pass.

```
Temporary test directory at /Volumes/tmp/test_runner_₿_🏃_20240402_192153
WARNING! There may be insufficient free space in /Volumes/tmp/test_runner_₿_🏃_20240402_192153 to run the Bitcoin functional test suite. Running the test suite with fewer than 2326.0 MB of free space might cause tests to fail.
```

Kept reducing the available space in the `/Volumes/tmp` partition, and a test eventually fails at 1.8 GB of free space

```
Filesystem      Size    Used   Avail Capacity iused ifree %iused  Mounted on
/dev/disk4s1   1.8Gi    12Mi   1.8Gi     1%      13  4.3G    0%   /Volumes/tmp
```

with the following error:

```
Early exiting after test failure due to disk running out of space in /Volumes/tmp/test_runner_₿_🏃_20240402_202655
```

### `--nocleanup`

Testing the with the `./test/functional/test_runner.py -j13 -t /Volumes/tmp --nocleanup` requirement for a minimum of 12 GB, with a partition of 11 GB and the `df -h` displaying:

```
Filesystem      Size    Used   Avail Capacity iused ifree %iused  Mounted on
/dev/disk4s2    11Gi    27Mi    11Gi     1%       4  4.3G    0%   /Volumes/tmp
```

I get the expected insufficient free space warning and all tests pass.

```
WARNING! There may be insufficient free space in /Volumes/tmp/test_runner_₿_🏃_20240402_203805 to run the functional test suite with --nocleanup. A minimum of 12 GB of free space is required.
```

Tested down to at 5.5 GB and eventually a test fails halting the test runner as well.

```
Early exiting after test failure due to disk running out of space in /Volumes/tmp/test_runner_₿_🏃_20240402_210436
```

File: test/functional/test_runner.py [hunk 11]
```diff
@@ -646,6 +667,9 @@ def run_tests(*, test_list, src_dir, build_dir, tmpdir, jobs=1, enable_coverage=
                     logging.debug("Early exiting after test failure")
                     break
 
+                if "[Errno 28] No space left on device" in stdout:
+                    sys.exit(f"Early exiting after test failure due to disk running out of space in {tmpdir}")
```
Comment (tdb3) [inline test/functional/test_runner.py:671]: nit:  The directory is left full upon exit.  The user would need to clean it up before re-running tests.  Having the user manually clean up the directory seems reasonable to me (the user may want to see actions leading to the failure).  An enhanced error message adds value.

```diff
diff --git a/test/functional/test_runner.py b/test/functional/test_runner.py
index dd3ea370bd..f08eb63b21 100755
--- a/test/functional/test_runner.py
+++ b/test/functional/test_runner.py
@@ -668,7 +668,9 @@ def run_tests(*, test_list, src_dir, build_dir, tmpdir, jobs=1, enable_coverage=
                     break
 
                 if "[Errno 28] No space left on device" in stdout:
-                    sys.exit(f"Early exiting after test failure due to disk running out of space in {tmpdir}")
+                    sys.exit(f"Early exiting after test failure due to insuffient free space in {tmpdir}\n"
+                             f"Test execution data left in {tmpdir}.\n"
+                             f"Additional storage is needed to execute testing.")
 
     print_results(test_results, max_len_name, (int(time.time() - start_time)))
 

```

**Review by tdb3 — COMMENTED** [review]
Nice work.  Adding free storage awareness is a great addition.

ACK for 2f7fcd5fe7289f7ad3efd566dbbd725264654f8f.
Left a nit inline.

Used a small ramdisk (1.2GB) with high jobs count (16) to force space exhaustion.  As expected, received the initial warning `Running the test suite with fewer than 2626.0 MB of free space might cause tests to fail.` and also received the exiting error `Early exiting after test failure due to disk running out of space...`

**Comment by BrandonOdiwuor** [comment]
Could you please take another look at this after incorporating the suggested changes

@achow101  @maflcko @kevkevinpal @davidgumberg @epiccurious @tdb3 @itornaza

**Comment by itornaza** [comment]
tested re-ACK for 858fa78637041ae704005d4b6e564dd8245f4b5d

## No size restrictions

- Did a code review and checked that there is adequate space on the file system to run the tests
- Configured with `--with-incompatible-bdb` and `--enable-suppress-external-warnings`
- Ran unit tests with `make check` and all tests pass
- Ran all functional tests with no extra flags, all tests pass and no warnings about free space are displayed
- Ran all functional tests with `--extended` and all tests pass as well

## Limited test space enforced using a ramdisk 

Ran the functional tests with the following options:
- `./test/functional/test_runner.py -t /Volumes/ramdisk` with  a ramdisk of 512 MB
- `./test/functional/test_runner.py -t /Volumes/ramdisk --nocleanup` with a ramdisk of 4 GB

The expected warnings are displayed upfront:

```
WARNING! There may be insufficient free space in /Volumes/ramdisk/test_runner_₿_🏃_20240411_211144 to run the Bitcoin functional test suite. Running the test suite with fewer than 1426.0 MB of free space might cause tests to fail.
```

and for the `--nocleanup` case:

```
WARNING! There may be insufficient free space in /Volumes/ramdisk/tmp/test_runner_₿_🏃_20240411_213828 to run the functional test suite with --nocleanup. A minimum of 12 GB of free space is required.
```

eventually when a test fails due to lack of disk space, the test harness exits and the new error message is displayed and is indeed much more informative!

```
Early exiting after test failure due to insuffient free space in /Volumes/ramdisk/tmp/test_runner_₿_🏃_20240411_212332
Test execution data left in /Volumes/ramdisk/tmp/test_runner_₿_🏃_20240411_212332.
Additional storage is needed to execute testing.
```

**Review by tdb3 — COMMENTED** [review]
re-ACK for 858fa78637041ae704005d4b6e564dd8245f4b5d

Thanks for incorporating the suggestion.

Used a small ramdisk (1.2GB) with high jobs count (20) to force space exhaustion. As expected, received the initial warning `Running the test suite with fewer than 3026.0 MB of free space might cause tests to fail`. and also received the exiting error `Early exiting after test failure due to disk running out of space...Test execution data left in ...Additional storage is needed to execute testing` with the new text added.

**Comment by davidgumberg** [comment]
reACK https://github.com/bitcoin/bitcoin/pull/29335/commits/858fa78637041ae704005d4b6e564dd8245f4b5d

Looks great, tested on a tmpfs 1 GiB in size, and the test runner prints the intended warning, and fails immediately when the disk is full.

File: test/functional/test_runner.py [hunk 12]
```diff
@@ -646,6 +667,11 @@ def run_tests(*, test_list, src_dir, build_dir, tmpdir, jobs=1, enable_coverage=
                     logging.debug("Early exiting after test failure")
                     break
 
+                if "[Errno 28] No space left on device" in stdout:
+                    sys.exit(f"Early exiting after test failure due to insuffient free space in {tmpdir}\n"
++                            f"Test execution data left in {tmpdir}.\n"
++                            f"Additional storage is needed to execute testing.")
```
Comment (achow101) [inline test/functional/test_runner.py:673]: In 858fa78637041ae704005d4b6e564dd8245f4b5d "test: Handle functional test disk-full error"

The `+`s at the beginning of the lines are unecessary.

```suggestion
                             f"Test execution data left in {tmpdir}.\n"
                             f"Additional storage is needed to execute testing.")
```

Comment (BrandonOdiwuor) [reply test/functional/test_runner.py]: fixed

**Comment by achow101** [comment]
ACK 7c0c599f1e6d412b3cb92cd67aa46644a8286cb6

**Review by cbergqvist — APPROVED** [review]
reACK 357ad110548d726021547d85b5b2bfcf3191d7e3. Looks good!

Tested on RAM disk of 1GB, 1430MB (`doas mount -t tmpfs -o size=1430m tmpfs /mnt/tmp/`) and varied the `--jobs` and `--nocleanup`, observing expected warnings + error on disk full.

**Review by tdb3 — COMMENTED** [review]
re ACK for 357ad110548d726021547d85b5b2bfcf3191d7e3

Retested https://github.com/bitcoin/bitcoin/pull/29335#pullrequestreview-1999297020 (and received the same results, which warn the user of insufficient free space and early exit due to insufficient space).

**Comment by itornaza** [comment]
re-ACK 357ad110548d726021547d85b5b2bfcf3191d7e3

Ran again all the tests as in https://github.com/bitcoin/bitcoin/pull/29335#issuecomment-2050304085 and I get the same results as expected on disk space in the cases it is being insufficient.

**Comment by achow101** [comment]
reACK 357ad110548d726021547d85b5b2bfcf3191d7e3
